### ISSUE TITLE
Flip `--experimental_sibling_repository_layout` to true

### ISSUE THREAD

**Issue by lberki** [bazelbuild/bazel#23576]
### Motivation

To make the location of a file in the execroot the same regardless of whether it's in an external repository or the main one.

### Description

Before the change, the source in the external repositories are located in the execroot at `$EXECROOT/<main repo>/external/<external repo>/<repo-relative path>`. This change turns that location into `$EXECROOT/<external repo>/<repo-relative-path>`.

Similarly, before the change, output files are located at `$EXECROOT/<main repo>/bazel-out/<config>/bin/external/<external repo>/<repo-relative path>` and this flag flip will turn that into `$EXECROOT/<main repo>/bazel-out/<external repo>/<config>/bin/<repo-relative path>`.

### Incompatible Flag

`--experimental_sibling_repository_layout`

### Migration Guide

This should mostly be a no-op change since everyone should refer to files using `file.path` in Starlark, `$(location <label)` in genrules and getting their names from the BEP when outside of Bazel.

If there are hard-coded paths in any of these places, they will need to be updated: preferably, to use one of these sanctioned approaches to learn the path or else hard-coding the new one, if that is not possible for some reason.

### In which Bazel LTS version will this incompatible change be enabled?

Bazel 8

**Comment by lberki** [comment]
@fmeum yes, I'm aware of all these downsides, especially the impending migration to bzlmod. Addressing these concerns in detail:

1. Yes, this is in fact a problem, but I see no way of providing a transition period where the files are available at both locations without either doubling the number of input files to actions (by providing them at both places) or teaching all the ways actions can be executed about sibling repository layout (and that includes all RBE implementations). So yes, it's painful, but it's a band-aid that should eventually be ripped off (unless, of course, we decide that it should never be done...)

2. Do you have evidence that the number of path segments is actually used? I would not be *very* surprised, but I was hoping that the use of `$(location)` and friends is prevalent enough that this would not be a big problem and our downstream projects pipeline seems to agree.

3. My game plan for this was that we'd move the repository-specific output directories to `../<external repo>/bazel-out/<config>/bin` _eventually_. I do realize that it's yet another incompatible change, but I thought that the value of reclaiming the `external` directory warrants it (the implementation of this would also not be simple, but that's our problem)

4. Yeah, I could argue both ways so I count this as an argument neither for nor against :)

@meteorcloudy mind weighing in?

**Comment by tjgq** [comment]
Would https://github.com/bazelbuild/bazel/pull/23518 completely address (1), or is there something else still missing?

**Comment by fmeum** [comment]
> Would https://github.com/bazelbuild/bazel/pull/23518 completely address (1), or is there something else still missing?

Not in its current form as it doesn't work for labels that produce multiple files. We may also need a way to get the relevant path for the current package without an explicit dependency on something (as required by location expansion).

There are a few related issues files on the Bazel repo that ideally would be addressed holistically.

**Comment by meteorcloudy** [comment]
@fmeum Thanks for the comprehensive analysis, it wasn't clear to me that all the breakages it would cause (most worried about breaking the runfiles library).

> What I wonder about in general: How important is reclaiming the external directory going to be for users?

I think this is a very important question to answer. @lberki IIRC, there is currently no large project being blocked on this? At this point, I feel the cost of this change might significantly outweigh the benefits.

If we do want to do it, we should probably come up with a better migration plan for all mentioned breakages, I'm not sure if we have enough time before Bazel 8 cut.

**Comment by meteorcloudy** [comment]
And is there any benefit from flipping `--legacy_external_runfiles` without flipping `--experimental_sibling_repository_layout`? It cannot free up the "external" directory alone, right.

**Comment by fmeum** [comment]
It can drastically cut down the number of symlinks in the sandbox, which is a performance win.

**Comment by lberki** [comment]
After discussing it with @meteorcloudy, I now think it's best to give up on this flag.

I do realize this is reminiscent of the discussion we had on https://github.com/bazelbuild/bazel/discussions/20500, where we came to the conclusion that it's best to flip the flag, but then, we thought that making the downstream projects pipeline green is good enough, which is in theory the case, but we can't ignore all the incompatibilities this change would cause that we don't know about :(

In addition, the new layout has shortcomings as @fmeum discussed above (there are good reasons why those shortcomings exist but they are shortcomings nevertheless) and those can only be solved by changing the layout, which would amount to an almost complete rewrite of the flag.

Further, whereas the fact that one cannot have a directory called `external/` in Bazel repositories is very embarrassing, I don't know anything that hurts a lot because of this limitation, so it's not that the high costs associated with the flag flip would bring a lot of benefit (other than removing an embarrassment).

So for this flag, I propose that we give up on it, make it a no-op, then, if we do find a good reason to try again, we do so with a better layout (probably `../<external repo>/blaze-out/<config>/bin`)

However, if we accept that this flag is best abandoned, it also calls into question the reason for `--legacy_external_runfiles`: for that, I propose that the right approach is to abandon the *new* locations instead of the old ones so that the layout of the execroot and the runfiles tree is consisten. Bazel cannot do that at the moment but it's trivial to teach it to. @meteorcloudy kindly agreed to run an experiment as to how harmful that would be.

I won't commit to this plan right now to give some room for protests, but given that the cost of flipping the flag is high (if done right) and that it's not perfect,  it's hard to imagine an argument...

**Comment by meteorcloudy** [comment]
> I propose that the right approach is to abandon the new locations instead of the old ones so that the layout of the execroot and the runfiles tree is consisten.

In case the new layout is already widely used (I do suspect so), I wonder if we can keep both layouts and reduce symlinks by doing:
```
bin.runfiles/
├── _main
│   ├── external
│   │   ├── repo1 -> ../../repo1
│   │   └── repo2 -> ../../repo2
│   └── pkg
│       └── bin
├── repo1
│   └── pkg
│       └── file
└── repo2
    └── pkg
        └── file
```

This will hopefully reduce the number of runfiles but keep things compatible.

**Comment by fmeum** [comment]
> I propose that the right approach is to abandon the new locations instead of the old ones so that the layout of the execroot and the runfiles tree is consisten.

I don't think that consistency between execroot and runfiles tree matters much to users. In fact, I actually consider seeming consistency to be dangerous: runfiles paths would only look like exec paths for source files, but would still deviate for generated files. With `--nolegacy_external_runfiles`, it's at least obvious that whatever works for source files will also work for generated files.

The non-external runfiles layout is also much nicer to work with in isolation as `../<other_repo>` gets you from any repo to any other. This is more relevant for runfiles, which are often used to run non-Bazelized tools that care about the exact layout of files rather than receiving their paths as arguments.

Even with `--experimental_sibling_repository_layout` abandoned in its current form, I think that flipping `--legacy_external_runfiles` to `False` is still the best option for runfiles and also results in a mostly manageable migration. It also makes sense to have this migration in parallel with the one to Bzlmod as there already are useful abstrations (`$(rlocationpath ...)`, runfiles libraries) that you can migrate to that help with both.

**Comment by lberki** [comment]
@fmeum I could be convinced either way (I'm not sure if the old or new locations are hard-wired more frequently), but I thought you specifically warned against flipping one flag but not the other at https://github.com/bazelbuild/bazel/issues/23574#issuecomment-2343429705 ?

I think the non-external layout is by far nicer to work with (just like for repositories), the question is just how much cost the flag flip incurs vs. how much nicer it is.

**Comment by fmeum** [comment]
I just wanted to call out additional breakage points. But these are already weak points (e.g. a generated Java runtime won't be found in runfiles with `$(JAVA)` even today). We could just add a new variable with the runfiles path (preferably even rlocationpath), add this to the migration notes and call it a day.

Just based on my experience, I would consider hardcoding of the new runfiles layout to be far more prevalent than of the old layout.

**Comment by lberki** [comment]
Ack, I thought that you think it's an especially bad breakage. I don't have data as to which approach would be less breaking, so I propose that we take a look at the Bazel downstream pipeline and make a decision based on the number of breakages there. @meteorcloudy WDYT?

**Comment by meteorcloudy** [comment]
> For us, these issues impact us more in the main workspace rather than external repos, but we certainly should try not to make paths longer.

Totally understand that every character on Windows is very precious. But I think that mostly apply to the MSVC compiler and you can find workaround for most cases? As @fmeum pointed out in previous discussions, the new sibling layout has a few drawbacks that are hard to deal with, not mention the migration cost. So I don't think this is a strong argument to push --experimental_sibling_repository_layout through.

**Comment by yuzhy8701** [comment]
One example that this flag is still useful is the "android-emulator-next" project ([link](https://cs.android.com/android/platform/superproject/+/emu-dev:build/bazel/base.bazelrc;drc=829c37323a82fc3482538503d8c8101c377dab33;l=6)), where we switched from cmake to bazel and rely on this flag to free up the `//external/...` namespace.

Generally speaking, this is the only benefit of the flag but an important one for projects where an `external` directory is already there and contains the majority of the codebase.

If this flag has to die, I would suggest providing some other means to access the `external` directory, preferably not having to declare everything as `new_local_repository`s?

**Comment by lberki** [comment]
@meteorcloudy annoying, but doable. The main problem with that is that it would necessarily have to be the decision made in the main repository, but one that would affect every other: say, if the main repository set `external` to be called `fluffykitten`, every project would have to say `-Ifluffykitten/project/include` instead of `-Iexternal/project/include`.

Sure, they should not hard-wire the string `external` in the first place, but if people did that, the migration to sibling repositories would be much easier.

**Comment by yuzhy8701** [comment]
> Sure, they should not hard-wire the string external in the first place, but if people did that, the migration to sibling repositories would be much easier.

One major reason why sibling repository breaks rules and builds is that, people rely on the [documented](https://bazel.build/remote/output-directories) directory layout when writing rules (e.g. extract a certain path fragment, assemble a path to a file, etc), and sibling repository changed that. The changes needed to make them work again with sibling repository is sometimes non-trivial (e.g. https://github.com/bazelbuild/bazel/pull/17276/commits/da9ef5bea3572741701d0c711fd6139699bde89a).

Simply changing the directory name will in a lot of cases not breaking that assumption unless they hardcode `external` - which is easy to fix by just `s/external/bazel_external_dirname()/g` (assume that builtin function returns the external directory name). For cases like `"-Iexternal/project/include"` you just do `"-I{}/project/include".format(bazel_external_dirname())` instead, which I suppose is a trivial fix?

Actually, you can even add a flag to set the name of that directory - so you'll never have a conflict again.

@lberki any thoughts?

**Comment by yuzhy8701** [comment]
> Sure, they should not hard-wire the string external in the first place, but if people did that, the migration to sibling repositories would be much easier.

The reason why rules and builds are broken due to sibling repository is that, people rely on the [documented](https://bazel.build/remote/output-directories) directory layout when they write the rules, but sibling repository changed that with per-repo `bazel-bin` directories, etc. Fixing those issues is sometimes non-trivial (e.g. https://github.com/bazelbuild/bazel/pull/17276/commits/da9ef5bea3572741701d0c711fd6139699bde89a).

In this regard, changing the `external` directory name is a much smaller change - a rule / build won't break unless they hardcode the `external` name. Then I suppose it would be easy to replace those hardcoded `external` with an API call that returns the external directory name (e.g. `-Iexternal/project/include` becomes `"-I{}/project/include".format(external_dirname())`).

You could even add a flag to set the directory name as user needs it, so no future conflict would ever happen again.

@lberki any thoughts?

**Comment by lberki** [comment]
@yuzhy8701 this is regrettable because this is the exact reason why I was originally so keen on flipping this flag: I couldn't have imagined that no one has a project that requires using the `external/` subdirectory :(

I don't think it's good idea to provide *two* mechanisms to piece together Bazel workspaces from multiple repositories because there is just too much room for adverse interaction between them and too much extra user interface.

FWIW, a large part of the problem with sibling repository layout is that people do hardcode the `external/` path segment and it's a word that's way too common to just simply search-and-replace for it. You'd also need to do this in your transitive dependencies...

So I'm afraid I don't have a very good alternative to you; in fact, `--experimental_sibling_repository_layout` originally came from the desire to support files under that directory.

@meteorcloudy @fmeum maybe you have a genius idea?

**Comment by fmeum** [comment]
No genius idea, but if we want to eventually support projects with an `external` directory, why don't we do a survey of the places in which users feel the need to hardcore the string? Then we can work on abstractions that resolve these use cases in other ways and eventually rename the directory or take another stab at consolidating the layout.

**Comment by lberki** [comment]
My model of the state of the world is that there aren't many reasons to harcode these things, people just do. I admit, though, that this is not based on any sort of data, it's only a suspicion.

**Comment by fmeum** [comment]
How about starting a discussion where we collect these cases? C++ rules are definitely a case where we need to offer a convenient alternative, but there could be more.

**Comment by peakschris** [comment]
> Actually, you can even add a flag to set the name of that directory as user needs it - no conflicts ever again

I like this idea. It seems flexible. With this flag, the experimental_sibling_repository_layout flag could be left unflipped and removed. Is there ever a reason to support sibling layout if the 'external' name can be changed to avoid conflict?

**Comment by lberki** [comment]
@peakschris see https://github.com/bazelbuild/bazel/issues/23576#issuecomment-2348577108 -- this is doable, but then every repository in the build would need to know what value the top-level project sets this string to. And if that's doable, we could just as well flip the flag as it is.

**Comment by meteorcloudy** [comment]
>  but then every repository in the build would need to know what value the top-level project sets this string to.

Or the root project just have to patch their dependencies for this renaming? I think for a project wants to work around this limitation, that's a fair price they need to pay? @yuzhy8701 Do you think this is acceptable?

**Comment by yuzhy8701** [comment]
@meteorcloudy Fair point.

> FWIW, a large part of the problem with sibling repository layout is that people do hardcode the `external/` path segment

That was my original point - if a project hardcodes `external/` it won't work with sibling repo and needs a fix anyways (if you flip the flag). I've made several fixes to some of our dependencies (rules_rust, grpc, and the bazel-internal cc rules). Compared with the fix for sibling repo, simply replacing `external` with something else is typically easier. This gives you a higher chance that a fix will ever be made, and that the main repo patching all its dependencies sound more viable when the fixes are easy.

**Comment by meteorcloudy** [comment]
@lberki OK, sounds like what we should do is 
- Add a flag for renaming the "external" directory (but never flip it). 
- Remove `--experimental_sibling_repository_layout` and its relevant logic.

**Comment by yuzhy8701** [comment]
SG - and also,

1. the API to access this value from where needed (in a backward compatible way)
2. the corresponding documentation (e.g. update https://bazel.build/remote/output-directories#layout-diagram to make people aware of the variable name).

**Comment by peakschris** [comment]
The renaming flag seems like a good approach. To elaborate on this one piece:

> 1. the API to access this value from where needed (in a backward compatible way)

We currently have 'external' hardcoded in two places:

a) In a run_binary (from aspect) rule, in both args and env sections. A variable substitution would be needed
b) Within a custom tool called by a bazel rule, the 'external' value would need to be passed to this as a command line input

**Comment by lberki** [comment]
@meteorcloudy if that's doable -- that string is embedded pretty deeply into Bazel and I'd not be surprised if some people even depended on the fact that the directory where external repositories are located in the output base (`$OUTPUT_BASE/external`) has the same basename as the directory under which they are located under the execroot.

I'll give it a stab as time allows to see how intrusive it would be. I hope I can make it a regular command line option and not a startup one.

**Comment by fmeum** [comment]
Since https://github.com/bazelbuild/bazel/pull/24126, non-main repos can now contain a top-level `external` directory or package even without this flag enabled.

Have projects that currently use the flag tried to bring in their `external` package as a separate repository via `local_repository`? Other than having to use `@external//pkg:foo` instead of `//external/pkg:foo`, I don't see any downside. Is there anything that makes this not a good idea in practice?

**Comment by lberki** [comment]
That approach has worked since ages immemorial, it's just that it's surprising and finicky.

**Comment by tpudlik** [comment]
Is the `local_repository` workaround not affected by https://github.com/bazelbuild/bazel/issues/4508 ?

**Comment by fmeum** [comment]
> Is the `local_repository` workaround not affected by https://github.com/bazelbuild/bazel/issues/4508 ?

The problem is a collision between the `external` subdirectory of the execroot and a main repo package called `external`. This does not apply to external repos, which are already placed under that `external` directory.

**Comment by keith** [comment]
https://github.com/bazelbuild/bazel/pull/24126 is a great workaround for general bzlmod modules, but if  `--experimental_sibling_repository_layout` goes away that means there could be modules that don't build at all on their own, and only build as a dependency via bzlmod, which feels a bit weird?

we're hitting this today with https://github.com/zeromq/libzmq which has a root `external/` directory

**Comment by fmeum** [comment]
A slightly odd idea that could get us `external` package support without a complex flag flip and costly migration:  Since we disabled legacy external runfiles and all Bzlnod repo names contain a `+`, we could allow all packages under `//external` whose first segment doesn't contain a `+` and be sure that they don't collide with external repos - they could live alongside each other in the execroot. In that case we should be a bit more careful about which characters we allow so that Bzlmod can still change its naming scheme if needed.

(Footnote: this would require moving the marker files to a subdirectory, but that's an internal change only)

@Wyverald @lberki Is this too much of a hack? But perhaps the alternatives (including the status quo) are worse?

**Comment by lberki** [comment]
Argh, I have been gearing up to _finally_ put a stake through the heart of `--experimental_sibling_repository_layout` :( I would be more comfortable deciding things if @Wyverald was around because he knows way more about the ecosystem than I do.

However, I do have an opinion about the proposed hack, which is that the situation is ugly enough that such a hack wouldn't make the situation any worse. That said, we'd have to not only disallow packages, but also files under `external/` since the top-level BUILD file can reference files like `//:external/foo+bar`, which AFAIU would also break this scheme. It would also make evaluating the top-level package slightly more expensive, but that is a pretty reasonable tradeoff, I think.

**Comment by meisterT** [comment]
This is obsolete: `--experimental_sibling_repository_layout` was made a no-op in ce2477d7eb8f1ca8f7f2b934b174820a0b271d47 (PR #30586).
